### PR TITLE
buildLuarocksPackage: accept structured luarocks config

### PR DESCRIPTION
**Pull request by teto**
There is an arbitrary mapping being done right now between nixpkgs lua infrastructre and luarocks config schema. This is confusing if you use lua so let's make it possible to use the lua names in the nixpkgs, thanks to the lib.generators.toLua convertor.

The only nixpkgs thing to remember should be to put the config into `luarocksConfig`

`buildLuarocksPackage.extraVariables` should become `buildLuarocksPackage.luarocksConfig.variables`

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/interpreters/lua-5/build-luarocks-package.nix [hunk 1]
```diff
@@ -116,23 +117,36 @@ let
...
-      '';
+
+    luarocksConfig' = lib.recursiveUpdate luarocksConfig
+      (lib.optionalAttrs (attrs ? extraVariables) (lib.warn "extraVariables in buildLuarocksPackage is deprecated, use luarocksConfig instead"
+      {
+        inherit (attrs) extraVariables;
+      }))
+    ;
+  in lib.recursiveUpdate generatedConfig luarocksConfig';
+
```
Comment (mrcjkb) [inline pkgs/development/interpreters/lua-5/build-luarocks-package.nix:149]: question: I'm a bit confused here, so I just want to make sure...

in `lua-modules/overrides.nix`, it looks like what used to be `extraVariables` is now passed in via `luarocksConfig.variables`.

But this looks to me like `extraVariables` are merged into `luarocksConfig`, not `luarocksConfig.variables`?

**Comment by mrcjkb** [comment]
question: With such a large amount of rebuilds, shouldn't this go into `staging`?

File: pkgs/development/interpreters/lua-5/build-luarocks-package.nix (hunk 1 above)
Comment (teto) [reply pkgs/development/interpreters/lua-5/build-luarocks-package.nix:149]: indeed !

**Comment by Izorkin** [comment]
@teto this PR breaks build of luadbi-mysql package:
```
lua5.2-luadbi-mysql> structuredAttrs is enabled
lua5.2-luadbi-mysql> Running phase: unpackPhase
lua5.2-luadbi-mysql> unpacking source archive /nix/store/npzmjakmxhqyshmr1m33yvb0y6faa9lf-luadbi-73a234c
lua5.2-luadbi-mysql> source root is luadbi-73a234c
lua5.2-luadbi-mysql> Running phase: patchPhase
lua5.2-luadbi-mysql> Running phase: updateAutotoolsGnuConfigScriptsPhase
lua5.2-luadbi-mysql> Running phase: configurePhase
lua5.2-luadbi-mysql> Running phase: buildPhase
lua5.2-luadbi-mysql> Running phase: installPhase
lua5.2-luadbi-mysql> Missing dependencies for luadbi-mysql 0.7.2-1:
lua5.2-luadbi-mysql>    luadbi 0.7.2 (not installed)
lua5.2-luadbi-mysql> luadbi-mysql 0.7.2-1 depends on lua >= 5.1, < 5.4 (5.2-1 provided by VM: success)
lua5.2-luadbi-mysql> luadbi-mysql 0.7.2-1 depends on luadbi 0.7.2 (not installed)
lua5.2-luadbi-mysql> mkdir: cannot create directory '/homeless-shelter': Permission denied
lua5.2-luadbi-mysql> Warning: Failed searching manifest: Failed downloading https://luarocks.org/manifest-5.2 - failed downloading https://luarocks.org/manifest-5.2
lua5.2-luadbi-mysql> Warning: Failed searching manifest: Failed downloading https://raw.githubusercontent.com/rocks-moonscript-org/moonrocks-mirror/master/manifest-5.2 - failed downloading https://raw.githubusercontent.com/rocks-moonscript-org/moonrocks-mirror/master/manifest-5.2
lua5.2-luadbi-mysql> Warning: Failed searching manifest: Failed downloading https://loadk.com/luarocks/manifest-5.2 - failed downloading https://loadk.com/luarocks/manifest-5.2
lua5.2-luadbi-mysql> Error: Could not satisfy dependency luadbi 0.7.2: No results matching query were found for Lua 5.2.

```

**Comment by teto** [comment]
the override was erasing the existing value, best to do a recursiveUpdate. As a quickfix you can do
```
diff --git a/pkgs/development/lua-modules/overrides.nix b/pkgs/development/lua-modules/overrides.nix
index 722c3d4940f0..7062fcbe404c 100644
--- a/pkgs/development/lua-modules/overrides.nix
+++ b/pkgs/development/lua-modules/overrides.nix
@@ -353,11 +353,13 @@ in
   });
 
   luadbi-mysql = prev.luadbi-mysql.overrideAttrs (oa: {
-    luarocksConfig.variables = {
-      # Can't just be /include and /lib, unfortunately needs the trailing 'mysql'
-      MYSQL_INCDIR = "${libmysqlclient.dev}/include/mysql";
-      MYSQL_LIBDIR = "${libmysqlclient}/lib/mysql";
-    };
+    luarocksConfig = oa.luarocksConfig // {
+        variables = {
+          # Can't just be /include and /lib, unfortunately needs the trailing 'mysql'
+          MYSQL_INCDIR = "${libmysqlclient.dev}/include/mysql";
+          MYSQL_LIBDIR = "${libmysqlclient}/lib/mysql";
+        };
+      };
     buildInputs = oa.buildInputs ++ [
       mariadb.client
       libmysqlclient
```

**Comment by Izorkin** [comment]
Thanks, patch works.

**Pull request merged**